lib/broadway_rabbitmq/producer.ex

```diff
@@ -437,18 +437,27 @@ defmodule BroadwayRabbitMQ.Producer do
       {_module, _channel, ack_data} = msg.acknowledger
 
       try do
-        case kind do
-          :successful -> apply_ack_func(ack_data.on_success, ack_data, channel)
-          :failed -> apply_ack_func(ack_data.on_failure, ack_data, channel)
+        case apply_ack_func(kind, ack_data, channel) do
+          {:error, reason} -> raise(reason)
+          success -> success
         end
       catch
         kind, reason ->
           formatted_error = Exception.format(kind, reason, __STACKTRACE__)
           Logger.error("Could not ack/reject message: #{formatted_error}")
+          :erlang.raise(kind, reason, __STACKTRACE__)
```

Comment:
I think we can get rid of the whole try/catch here. The biggest question, however, is what happens with the other messages? I think we should try to ack everything possible and then raise only if there is something we could not ack.  Something like this:

```elixir
    Enum.flat_map(messages, fn msg ->
      {_module, _channel, ack_data} = msg.acknowledger

      case apply_ack_func(kind, ack_data, channel) do
        {:error, reason} -> [{msg, reason}]
        {:ok, _} -> []
      end
    end)
    |> case do
      [] -> :ok
      failed -> raise ...
    end
```

WDYT?

Reply:
I think it's a good idea! It seems the only reason that it's there is to create a log entry.

I'll drop it in and push once I confirm the test suite is green.